### ISSUE TITLE
Ray-Sphere intersection test is not numerically robust

### ISSUE THREAD

**Issue by TzuChieh** [TzuChieh/Photon-v2#7]
Rendering spherical area light with tiny radius (~0.005) starts to show signs of numeric error. Specifically, render the following scene with triangles then with a sphere produces different results:

```plaintest
## camera(pinhole) [real fov-degree 30] [vector3r position "0 6 40"] [vector3r direction "0 0 -1"]
## film(hdr-rgb) [integer width 800] [integer height 800] [string filter-name "gaussian"]
## sample-generator(stratified) [integer sample-amount 10000][integer num-strata-2d-x 800][integer num-strata-2d-y 800]
## integrator(bneept)
//## integrator(surface-normal)

-> geometry(rectangle) @plane [real width 15] [real height 15]
-> geometry(rectangle) @smallPlane [real width 0.5] [real height 0.5]
-> geometry(sphere)    @smallBall [real radius 0.005]
//-> geometry(sphere)    @target  [real radius 3]
-> geometry(rectangle) @target [real width 4] [real height 4]

-> material(matte-opaque) @white [vector3r albedo "0.8 0.8 0.8"]
//-> material(abraded-opaque) @white [vector3r albedo "0.8 0.8 0.8"] [real roughness-v 0.02] [real roughness-u 0.1] [vector3r f0 "0.9 0.9 0.9"]

-> actor(model) @ground [geometry geometry @plane] [material material @white]
-> actor(model) rotate(@ground) [vector3r axis "1 0 0"] [real degree -90]
-> actor(model) scale(@ground) [vector3r factor "10 10 10"]

-> actor(model) @object [geometry geometry @target]  [material material @white]
-> actor(model) translate(@object) [vector3r factor "0 4 0"]
-> actor(model) rotate(@object) [vector3r axis "1 0 0"] [real degree -90]

-> light-source(area) @areaSource [vector3r emitted-radiance "400000 400000 400000"]
-> actor(light) @topLight [geometry geometry @smallBall] [material material @white] [light-source light-source @areaSource]
-> actor(light) translate(@topLight) [vector3r factor "0 10 0"]
-> actor(light) rotate(@topLight) [vector3r axis "1 0 0"] [real degree 90]
-> actor(light) rotate(@topLight) [vector3r axis "0 0 1"] [real degree 45]
```

Currently, I believe the problem lies in ray-sphere intersection test.

**Comment by TzuChieh** [comment]
The commit regarding this issue:
e96513ea0be5d96801a66016720eeae0baf3da9d

**Comment by TzuChieh** [comment]
You can also see the problem by rendering this scene with triangulated sphere and then true sphere:

```plaintext
## camera(pinhole) [real fov-degree 30] [vector3r position "0 3 6"] [vector3r direction "0 -0.35 -1"]
## film(hdr-rgb) [integer width 800] [integer height 800] [string filter-name "gaussian"]
## sample-generator(stratified) [integer sample-amount 400][integer num-strata-2d-x 800][integer num-strata-2d-y 800]
## integrator(bneept)

-> geometry(rectangle) @plane [real width 15] [real height 15]
-> geometry(rectangle) @smallPlane [real width 0.5] [real height 0.5]
-> geometry(sphere)    @smallBall [real radius 0.005]
-> geometry(rectangle) @target [real width 1] [real height 1]

-> material(matte-opaque) @white [vector3r albedo "0.8 0.8 0.8"]
//-> material(abraded-opaque) @white [vector3r albedo "0.8 0.8 0.8"] [real roughness-v 0.02] [real roughness-u 0.1] [vector3r f0 "0.9 0.9 0.9"]

-> actor(model) @ground [geometry geometry @plane] [material material @white]
-> actor(model) rotate(@ground) [vector3r axis "1 0 0"] [real degree -90]
-> actor(model) scale(@ground) [vector3r factor "10 10 10"]

-> actor(model) @object [geometry geometry @target]  [material material @white]
-> actor(model) translate(@object) [vector3r factor "0 1 0"]
-> actor(model) rotate(@object) [vector3r axis "1 0 0"] [real degree -90]

-> light-source(area) @areaSource [vector3r emitted-radiance "40000 40000 40000"]
-> actor(light) @topLight [geometry geometry @smallBall] [material material @white] [light-source light-source @areaSource]
-> actor(light) translate(@topLight) [vector3r factor "0 2 0"]
-> actor(light) rotate(@topLight) [vector3r axis "1 0 0"] [real degree 90]
-> actor(light) rotate(@topLight) [vector3r axis "0 0 1"] [real degree 45]
```

**Comment by TzuChieh** [comment]
bub fixed: 6d13ccb2728fac4a9f9ae4656fa7ca98e8b47bf3
Still, the ray-sphere intersection test's robustness is yet to be tested.
